Correct datatype of name and is_memory_optimized columns in sys.tables

timchang514 opened the pull request:
As part of import-export support, we need case-insensitive matching for the
name column of sys.tables. We can apply a similar fix here as was used for
information_schema.tables by using a correctly collated copy of sys.sysname,
until sysname collation issues can be sorted out. Also, the is_memory_optimized
column should be a sys.bit.

Signed-off-by: Tim Chang <changti@amazon.com>

### Description

Correcting two column datatypes in sys.tables, name and is_memory_optimized. 
 
### Issues Resolved

These two changes are required for supporting Import-Export Wizard with Babelfish as a source. 


### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

timchang514 committed:
Correct datatype of name and is_memory_optimized columns in sys.tables

As part of import-export support, we need case-insensitive matching for the
name column of sys.tables. We can apply a similar fix here as was used for
information_schema.tables by using a correctly collated copy of sys.sysname,
until sysname collation issues can be sorted out. Also, the is_memory_optimized
column should be a sys.bit.

Signed-off-by: Tim Chang <changti@amazon.com>

2jungkook committed:
Bumping up versions and adding upgrade scripts (#129)

* Upgrade scripts for both major and minor version upgrades

Signed-off-by: Jungkook Lee <jungkook@amazon.com>

* Fix typo

Signed-off-by: Jungkook Lee <jungkook@amazon.com>

Co-authored-by: Jungkook Lee <jungkook@amazon.com>

thephantomthief committed:
Reverted "Special argument handling in suser_ functions (#708)"

Commit 248885df2f51df9cba789ddf59cced9a40ccca01 introduced some upgrade
failures so reverted the same.

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief committed:
Add Github Action for running JDBC tests by upgrading Babelfish extensions

This commit adds Github Action to test the upgrade workflow for
Babelfish. The action does the following:
1. Builds the postgres engine modified for babelfish of version 14
2. Builds and installs the babelfish extensions with babelfishpg_tsql extension
   version being 2.0.0
3. Initializes the babelfish database
4. Builds the postgres engine modified for babelfish of current dev
   version
5. Builds and installs the babelfish extensions with babelfishpg_tsql extension
   version being 2.1.0
6. Updates the babelfish extensions using ALTER EXTENSION ... UPDATE
7. Runs the JDBC tests

Signed-off-by: Sharu Goel <goelshar@amazon.com>

thephantomthief reviewed (changes requested):
Please add upgrade script changes and test cases.

rishabhtanwar29 committed:
Fix minor-version-upgrade github action

While building the new extensions binary, we were checking out the
wrong repository; because of that, the changes introduced in the PR were
not getting tested. Fixed the same.

Signed-off-by: Sharu Goel <goelshar@amazon.com>

HarshLunagariya committed:
Update lock_timeout in fk-deadlock isolation test

Previously, this test was failing while testing with remote instance because
lock_timeout was set to 100ms which was not sufficient for running them on
an instance.

This commit increases lock_timeout for session s1 to 500ms from 100ms.

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

Deepesh125 committed:
Some improvements in TDS fault injection framework

This commit fixes the following issues as described below:

1. Freeing string info was not correctly accounted in a few fault injection callback functions like tamper_rpc_request and tamper_request_sequential. Which was causing the server to crash. So, this change fixes this issue. Also, We should not reset Message memory context while reading and processing the request. So fixed that as well.

2. Certain types of fault injection should be called for certain type of requests. For example, tamper_rpc_request is supposed to be called for RPC request. This commit fixes this by checking request type inside fault injection callback before trying to tamper the request.

3. We were making use of list_delete() API incorrectly to disable any fault injection trigger inside FaultInjectionDisableTest(). Which was causing error like “unrecognised node type: 1600483952” to occur since equal() function was not able to find the node type. So this commit fixes that issue by using list_delete_cell() API to delete certain node from injected entries.

4. For "tds_comm_throw_error" trigger, we were throwing ERROR while processing the header (while reading the request). Any error raising from such a low level should be FATAL. So, introduced throw_error_comm() callback for "tds_comm_throw_error" trigger which will raise the FATAL error while processing the header if fault injection trigger is enabled.

5. Added TDS fault injection tests as well.

Task: BABEL-3071
Signed-off-by: Dipesh Dhameliya (dddhamel@amazon.com)

Deepesh125 committed:
Use logical schema name to write table name in column metadata instead of physical schema name.

Previously, Physical schema name i.e., master_dbo, temp_dbo, etc was being written as part of column metadata while sending the response.But logical schema should be written as far as Babelfish concerned.

In this commit, I introduced the rendezvous variable pltsql_get_logical_schema_name which would be initialised by babelfishpg_tsql extension to get_logical_schema_name_from_physical_schema_name().

get_logical_schema_name_from_physical_schema_name() would first check if given physical schema name needs to be translated to logical schema name or not. And it uses existing api get_logical_schema_name() to retrieve the logical schema name given physical schema name if required.

Task: BABEL-2890
Signed-off-by: Dipesh Dhameliya (dddhamel@amazon.com)

Deepesh125 committed:
Configure correct sortId for the collations supported in Babelfish

Previously, sortId field for the most of collations was configured incorrectly. Any non-zero value of sortId might cause
some client to choose different client encoding than the information provided in the lcid of the collation metadata. Due
to this, client might not able to parse some chars correctly and would show them incorrectly to the end user even though
data is correct byte wise.

This commit fixes the issue by configuring sortId to zero (default sort order) for the most of the collations. Hence,
client would use the lcid to determine the client encoding and would show the data correctly to end user.

Task: BABEL-3130
Signed-off-by: Dipesh Dhameliya dddhamel@amazon.com

Deepesh125 committed:
Introducing japanese collations

Adding set of Japanese collations Japanese_CS_AS, Japanese_CI_AI and Japanese_CI_AS. Also added a few test cases to verify the behaviour.

Task: BABEL-2942
Signed-off-by: Dipesh Dhameliya dddhamel@amazon.com

KushaalShroff committed:
Add Support for TDS Column-metadata flags

Earlier we were sending generic flags for each of the columns.
With this commit we have added support to send computed and identity
flags for their respective columns.

Task: BABEL-217
Signed-off-by: Kushaal Shroff (kushaal@amazon.com)

KushaalShroff committed:
Re-do the implementation done for sp_describe_first_result_set

Previously sp_describe_first_result_set procedure makes use of sp_describe_first_result_set_internal function which returned wrong values.

With this commit we have introduced a C function ‘sp_describe_first_result_set_internal’ which creates a view using the TSQL query provided by User and then uses this view to gain the required information from the existing system views. For now only Select queries are supported since it unblocks BCP Utility. With some basic testing, we are now able to perform ‘BCP IN’ and ‘BCP OUT’.

Added test cases to JDBC Framework.

Task: BABEL-3000
Signed-off-by: Kushaal Shroff (kushaal@amazon.com)

shalinilohia50 committed:
Additional row in sp_statistics_100 for heap table

In Tsql, sp_statistics_100 outputs an extra row
with TYPE=0 when a table has no clustered index. This
will always be the case in Babelfish, but sp_statistics_100
currently does not output that row.

Task: BABEL-3011
Authored-by: Shalini Lohia lshalini@amazon.com
Signed-off-by: Kuntal Ghosh kuntalgh@amazon.com

Deepesh125 committed:
Use logical schema name to write table name in column metadata instead of physical schema name

Previously, we introduced rendezvous variable pltsql_get_logical_schema_name babelfishpg_tsql extension to get_logical_schema_name_from_physical_schema_name().
    
With these changes, we are assuming that we must have received the valid schema name from engine while sending column metadata on wire. So updated the rendezvous variable pltsql_get_logical_schema_name to initialize to get_logical_schema_name(). We first try to get the logical schema name given physical schema name; and if we could not find the logical schema name then we can say that received schema name is shared schema and it does not need to be translated to logical schema name.

Task: BABEL-2890
Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>

rishabhtanwar29 committed:
Remove unused CASTS from pg_catalog.BOOL to sys.bpchar and sys.varchar

The CASTS from pg_catalog.BOOL to sys.bpchar and sys.varchar depend
upon certain implicit CASTS due to which major version upgrade from PG 13
to 14 fails with dependency issue.

To fix this, this commit completely removes these two problematic CASTS.
These CASTS are never used since bool is not a valid T-SQL datatype.

Task: BABEL-3171
Authored-by: Rishabh Tanwar <ritanwar@amazon.com>
Signed-off-by: Kuntal Ghosh kuntalgh@amazon.com

xhfanhe committed:
Support database-level user-defined roles - wave 1

Previously database-level user-defined roles are not supported in
Babelfish.

This commit includes implementation for
1. sys.database_role_members view
2. CREATE/ALTER/DROP ROLE commands with basic functionality
3. Role renaming - ALTER ROLE...WITH NAME...
4. Support in all related ownership objects

Note that this commit doesn't include CREATE ROLE WITH AUTHORIZATION.
It will be implemented in future commits.

Task: BABEL-3080/3081/3083
Signed-off-by: Xiaohui Fanhe <hexiaohu@amazon.com>

timchang514 committed:
Correct datatype of name and is_memory_optimized columns in sys.tables

As part of import-export support, we need case-insensitive matching for the
name column of sys.tables. We can apply a similar fix here as was used for
information_schema.tables by using a correctly collated copy of sys.sysname,
until sysname collation issues can be sorted out. Also, the is_memory_optimized
column should be a sys.bit.

Signed-off-by: Tim Chang <changti@amazon.com>
